pyperformance/data-files/benchmarks/bm_base64/run_benchmark.py

```diff
@@ -0,0 +1,274 @@
+"""Benchmark for the base64 module's primary public APIs.
+
+Tests encoding and decoding performance across various variants
+and data sizes, split into _small (balanced small data) and _large variants.
+"""
+
+import base64
+import random
+import pyperf
+
+
+# Generate test data with fixed seed for reproducibility
+random.seed(12345)
+DATA_TINY = bytes(random.randrange(256) for _ in range(20))
+DATA_SMALL = bytes(random.randrange(256) for _ in range(127))  # odd on purpose
+DATA_MEDIUM = bytes(random.randrange(256) for _ in range(3072))
+DATA_9K = bytes(random.randrange(256) for _ in range(9000))
+DATA_LARGE = bytes(random.randrange(256) for _ in range(102400))
+DATA_HUGE = bytes(random.randrange(256) for _ in range(1048576))
+
+# Pre-encoded data for decode benchmarks
+B64_TINY = base64.b64encode(DATA_TINY)
+B64_SMALL = base64.b64encode(DATA_SMALL)
+B64_MEDIUM = base64.b64encode(DATA_MEDIUM)
+B64_9K = base64.b64encode(DATA_9K)
+B64_LARGE = base64.b64encode(DATA_LARGE)
+B64_HUGE = base64.b64encode(DATA_HUGE)
+
+B64_URLSAFE_TINY = base64.urlsafe_b64encode(DATA_TINY)
+B64_URLSAFE_SMALL = base64.urlsafe_b64encode(DATA_SMALL)
+B64_URLSAFE_MEDIUM = base64.urlsafe_b64encode(DATA_MEDIUM)
+B64_URLSAFE_9K = base64.urlsafe_b64encode(DATA_9K)
+
+B32_TINY = base64.b32encode(DATA_TINY)
+B32_SMALL = base64.b32encode(DATA_SMALL)
+B32_MEDIUM = base64.b32encode(DATA_MEDIUM)
+B32_9K = base64.b32encode(DATA_9K)
+B32_LARGE = base64.b32encode(DATA_LARGE)
+B32_HUGE = base64.b32encode(DATA_HUGE)
+
+B16_TINY = base64.b16encode(DATA_TINY)
+B16_SMALL = base64.b16encode(DATA_SMALL)
+B16_MEDIUM = base64.b16encode(DATA_MEDIUM)
+B16_9K = base64.b16encode(DATA_9K)
+B16_LARGE = base64.b16encode(DATA_LARGE)
+B16_HUGE = base64.b16encode(DATA_HUGE)
+
+A85_TINY = base64.a85encode(DATA_TINY)
+A85_SMALL = base64.a85encode(DATA_SMALL)
+A85_MEDIUM = base64.a85encode(DATA_MEDIUM)
+A85_9K = base64.a85encode(DATA_9K)
+A85_LARGE = base64.a85encode(DATA_LARGE)
+A85_HUGE = base64.a85encode(DATA_HUGE)
+
+B85_TINY = base64.b85encode(DATA_TINY)
+B85_SMALL = base64.b85encode(DATA_SMALL)
+B85_MEDIUM = base64.b85encode(DATA_MEDIUM)
+B85_9K = base64.b85encode(DATA_9K)
+B85_LARGE = base64.b85encode(DATA_LARGE)
+B85_HUGE = base64.b85encode(DATA_HUGE)
+
+
+# --- Base64 (includes validate=True) ---
+
+def bench_b64_small(loops):
+    range_it = range(loops)
+    t0 = pyperf.perf_counter()
+    for _ in range_it:
+        for _ in range(450):
+            base64.b64encode(DATA_TINY)
+            base64.b64decode(B64_TINY)
+            base64.b64decode(B64_TINY, validate=True)
```

Comment:
Decoding has twice larger weight than encoding. `validate=True` should not affect performance for such input in reasonable implementation.

On other hand, I would test decoding from an ASCII string.

If you keep several decoding calls, you should balance them by equal number of encoding calls.